tarpc/src/client.rs

```diff
@@ -352,6 +352,7 @@ where
                     let _entered = span.enter();
                     tracing::info!("ReceiveError");
                 }
+                self.pending_requests_mut().close();
```

Comment:
The thing that's confusing me is, pending requests should close once the request dispatch is dropped. Can you should me your code that polls the request dispatch? What does the code do after `RequestDispatch::poll` returns an error?

Reply:
According to the [documentation](https://docs.rs/tokio/latest/tokio/sync/mpsc/index.html#clean-shutdown) of `tokio` it is possible to have messages in the channel after dropping the `Receiver`. This does not happen on each run but maybe once in 10 or 100 times.  My test currently produces the following log.
```
2024-01-23T22:14:29.394867Z  INFO tarpc::client: dropping request dispatch
2024-01-23T22:14:29.394871Z  INFO RPC{rpc.deadline=2024-01-23T22:14:39.394769825Z otel.kind="client" otel.name="TarpcCspVault.idkg_gen_dealing_encryption_key_pair" rpc.trace_id=00}: tarpc::client: sending request to dispatch
2024-01-23T22:14:29.394973Z TRACE tarpc::server: Expired requests: Closed, Inbound: Closed
2024-01-23T22:14:29.395005Z TRACE tarpc::server: poll_flush
2024-01-23T22:14:29.395019Z TRACE tokio_util::codec::framed_impl: flushing framed transport
2024-01-23T22:14:29.395032Z TRACE tokio_util::codec::framed_impl: framed transport flushed
2024-01-23T22:14:29.395025Z  WARN tarpc::client: Connection broken: could not read from the transport

Caused by:
    frame size too big
2024-01-23T22:14:29.395086Z  INFO RPC{rpc.deadline=2024-01-23T22:14:39.394769825Z otel.kind="client" otel.name="TarpcCspVault.idkg_gen_dealing_encryption_key_pair" rpc.trace_id=00}: tarpc::client: sending request to dispatch done; waiting for response
test rpc_connection::should_unfortunately_be_dead_after_response_from_server_cannot_be_received_by_client_because_too_large has been running for over 60 seconds
```

Our code makes use of the `tarpc` `Client` as a black box by calling its generated methods using `tokio`'s `block_on`. Here's an [example](https://github.com/dfinity/ic/blob/master/rs/crypto/internal/crypto_service_provider/src/vault/remote_csp_vault/tarpc_csp_vault_client.rs#L271).

Reply:
Thanks! Will review your code and get back to you.

(It shouldn't be a problem if there are messages in the channel when the channel is dropped — the messages in the channel contain the sender side of oneshot channels used to send individual RPC responses. When the channel is dropped, the senders are also dropped, and that should cause the receiver side to receive an error.)

Reply:
Thank you!

I'm not sure if the channel is actually dropped. The only thing that is dropped (IIUC) is the receiver part of the dispatch channel and some messages from the channel. But the channel itself, again IIUC, is just a synchronized queue that is shared by both sender and receiver. The sender part is not dropped until we drop the client, right, it only changes to an errored state to prevent sending further messages to the channel? My hypothesis so far was that:
1) A dispatch request (which contains the sender part of the response channel) is sent to the request dispatcher.
2) The caller waits on the receiver end of the response channel.
3) The dispatcher receiver is being dropped but the channel is not properly cleared, so the dispatch request along with the response sender handle is not dropped.
4) Client waits indefinitely.

WDYT @tikue?

Reply:
Ah, that's really interesting and makes a lot of sense, thanks!